### PR TITLE
Fix defects in code-coverage.gralde to generate code coverage report properly

### LINKED ISSUE #1213: [BUG] Code coverage report can't be generated properly due to defect in code-coverage.gradle
**Describe the bug**
There are 2 bugs in `code-coverage.gradle`:
1. When running `./gradlew check -Dtests.coverage=true`, the following error occurs:
```
Suite: Test class org.opensearch.test.loggerusage.OpenSearchLoggerUsageTests
  2> java.lang.RuntimeException: unable to install test security manager
        at org.opensearch.bootstrap.BootstrapForTesting.<clinit>(BootstrapForTesting.java:190)
        at org.opensearch.test.OpenSearchTestCase.<clinit>(OpenSearchTestCase.java:251)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:468)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$2.run(RandomizedRunner.java:623)

        Caused by:
        java.lang.NullPointerException: Cannot invoke "String.length()" because "input" is null
...
```
2. Whether the system property ["tests.coverage.html_report"](https://github.com/opensearch-project/OpenSearch/blob/83332c8ab6f8c134b09b9a3dc57ddf39a507859e/gradle/code-coverage.gradle#L83) is set or not, the html format code coverage report is always generated.

Reason:
1. I used the same System property "tests.coverage" from an old [code](https://github.com/opensearch-project/OpenSearch/blob/0ba0e7cc26060f964fcbf6ee45bae53b3a9941d0/test/framework/src/main/java/org/opensearch/bootstrap/BootstrapForTesting.java#L132) into my [gradle script](https://github.com/opensearch-project/OpenSearch/blob/83332c8ab6f8c134b09b9a3dc57ddf39a507859e/gradle/code-coverage.gradle#L89).
2. html format code coverage report is generated by default by JaCoCo gralde plugin, see [here](https://github.com/gradle/gradle/blob/4fa9149b3148554bc84c2f6e12ec7b53c10e4e2f/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java#L159)

**To Reproduce**
For bug 1: run `./gradlew check -Dtests.coverage=true`
For bug 2: run `./gradlew codeCoverageReportForUnitTest` or `./gradlew codeCoverageReport`, without having `-Dtests.coverage.html_report`, the code coverage report in html format is generated in `/build/reports/jacoco/`

**Expected behavior**
1. `./gradlew check -Dtests.coverage=true` can run gradle check and generate code coverage report without error.
2. Without having `tests.coverage.html_report` property, html format shouldn't be generated, and only xml format is generated. Only when adding the system property `tests.coverage.html_report`, such as running `./gradlew codeCoverageReport -Dtests.coverage.html_report`, the report in html format can be generated.

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**
Imported by https://github.com/opensearch-project/OpenSearch/pull/971
Related issue: https://github.com/opensearch-project/OpenSearch/issues/850

### PR DESCRIPTION
**Pull request by tlfeng**
### Description
- Fix the issue that code coverage in HTML format can't be disabled.
- Resolve the error when running `./gradlew check -Dtests.coverage=true`
 
Explanation:
1. Html format report is generated by default in [JaCoCo gradle plugin](https://github.com/gradle/gradle/blob/4fa9149b3148554bc84c2f6e12ec7b53c10e4e2f/subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java#L159), so it needs to be disabled manually.
2. System property `tests.coverage` is introduced in https://github.com/elastic/elasticsearch/commit/2a129f186874689f7c16b8ab77eba508d9ae7250, then I used the same property for generating the code coverage report, and it caused the error (a past forum thread about the error https://discuss.elastic.co/t/estestcase-2-x-jacoco-not-working/45104).
- I think that part of code is out dated, as the build system was changed from at that time, and the jacoco file doesn't have permission issue now when Java security manager is enabled. And Java security manager is also [being deprecated](https://github.com/opensearch-project/OpenSearch/issues/528#issuecomment-824787677). [@rmuir Please correct me if I'm wrong. Thank you!]

Future plan:
Considering refactoring the groovy script `code-coverage.gradle` to Java code and put into `/buildSrc` (see [here](https://docs.gradle.org/6.6.1/userguide/organizing_gradle_projects.html#sec:build_sources) to learn more about "buildSrc" directory) , then unit test can be written to test the build script.

### Issues Resolved
Resolves #1213 
 
### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
How is this used?

If this is used with `tests.coverage.html_report=false/true`, then the code should translate that value and enable/disable the report accordingly.

**Comment by tlfeng** [comment]
> How and where is this used?
> 
> If this is used with `tests.coverage.html_report=false/true`, then the code should translate that value and enable/disable the report accordingly.

Hi @dblock , thanks for your review.
The code coverage report has not been used yet, and I'm thinking of generating and uploading the report through CI workflow.
About the usage of `tests.coverage.html_report`, my design is to use with:
 - `./gradlew codeCoverageReport -Dtests.coverage.html_report` when report in HTML format is needed,
 - `./gradlew codeCoverageReport` when report in HTML format is not needed.
The value of `tests.coverage.html_report` doesn't matter, and whenever there is a property named `tests.coverage.html_report`, the HTML report will be generated.

**Comment by dblock** [comment]
> > How and where is this used?
> > If this is used with `tests.coverage.html_report=false/true`, then the code should translate that value and enable/disable the report accordingly.
> 
> Hi @dblock , thanks for your review.
> The code coverage report has not been used yet, and I'm thinking of generating and uploading the report through CI workflow.
> About the usage of `tests.coverage.html_report`, my design is to use with:
> 
> * `./gradlew codeCoverageReport -Dtests.coverage.html_report` when report in HTML format is needed,
> * `./gradlew codeCoverageReport` when report in HTML format is not needed.
>   The value of `tests.coverage.html_report` doesn't matter, and whenever there is a property named `tests.coverage.html_report`, the HTML report will be generated.
> 
> I could modify the usage if you and others have other opinions by this chance. 😁

Sounds like you want `-Dtests.coverage.report=html` or `-Dtests.coverage.report=txt` or ``-Dtests.coverage.report=html,txt` - wouldn't that be clearer?

**Comment by tlfeng** [comment]
> Sounds like you want `-Dtests.coverage.report=html` or `-Dtests.coverage.report=txt` or ``-Dtests.coverage.report=html,txt` - wouldn't that be clearer?

@dblock Thanks for your your idea! It looks clearer, but make the command to get the report more complex.
There are 3 kinds of report can be generated, `xml`, `html` and `csv`.
1. `xml` format can be used by code coverage analysis software, like Codecov that we are using.
2. `html` format is human readable, and will only used for people to read the report.
3. I have no idea with `csv` format report, and has never been used so far.

By default, `xml` and `html` format will be generated by JaCoCo gradle plugin.
The idea to make `html` format optional is it takes 100MB+ disk space, and when used in CI workflow, nobody will read the report, and only `xml` format will be uploaded to Codecov for analysis.
So I wanted to have a parameter to disable the report in HTML format, and wanted to have report in XML format all the time when collecting the code coverage.

Anyway, changing the parameter to `-Dtests.coverage.report=html` is reasonable, and I could change this.

**Comment by tlfeng** [comment]
I renamed the parameter to control the format of the code coverage report.
By default, code coverage report in XML format will be generated.
If want additional HTML format, adding the System property "tests.coverage.report.html" by appending `-Dtests.coverage.report.html`.
If want additional CSV format, adding the System property "tests.coverage.report.html" by appending `-Dtests.coverage.report.csv`.

File: gradle/code-coverage.gradle [hunk 1]
```diff
@@ -79,10 +79,11 @@ tasks.withType(JacocoReport).configureEach {
...
     xml.enabled true
-    // Code coverage report in HTML format is on demand， since it takes up 100+MB of disk space.
-    if (System.getProperty('tests.coverage.html_report')) {
-      html.enabled true
-    }
+    // Code coverage report in HTML and CSV formats are on demand, in case they take extra disk space.
+    if (!System.getProperty('tests.coverage.report.html'))
+      html.enabled false
+    if (System.getProperty('tests.coverage.report.csv'))
+      csv.enabled true
```
Comment (dblock) [inline gradle/code-coverage.gradle:86]: The comment says that these are on-demand, but the code tells me that by default `html.enabled=true` and `csv.enabled=false`. Where are those defaults? 

I'd expect this to be something like this:

```
xml.enabled System.getProperty('tests.coverage.report.xml', true)
html.enabled System.getProperty('tests.coverage.report.html', false)
csv.enabled System.getProperty('tests.coverage.report.html', false)
```

**Review by dblock — CHANGES_REQUESTED** [review]
Question:

File: gradle/code-coverage.gradle (hunk 1 above)
Comment (tlfeng) [reply gradle/code-coverage.gradle:86]: Ah, you got a great idea! Thank you! It makes the code much clearer. 😄

In addition, if the "reports" configurations are not explicitly written, report in `xml` and `html` format will be generated, so it means this is the default behavior for "JaCoCo Gradle plugin".
```
reports {
    xml.enabled ...
    html.enabled ...
    csv.enabled ...
}
```

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
start gradle check

**Review by dblock — APPROVED** [review]

**Pull request merged**